src/analyzer/protocol/ntp/ntp-protocol.pac

```diff
@@ -0,0 +1,124 @@
+	version	:	uint8  = (first_byte & 0x38)>>3; 	# Bits 3-5 of 8-bits value
+	mode	:	uint8  = (first_byte & 0x07);  		# Bits 6-8 of 8-bits value
+} &byteorder=bigendian &exportsourcedata;
+
+# This is the most common type of message, corresponding to modes 1-5
+# This kind of msg are used for normal operation of syncronization
+# See RFC 5905 for details
+type NTP_std_msg = record {
+        stratum        : uint8;
+        poll           : int8;
+        precision      : int8;
+
+        root_delay     : NTP_Short_Time;
+        root_dispersion: NTP_Short_Time;
+        reference_id   : bytestring &length=4;
+        reference_ts   : NTP_Time;
+
+        origin_ts      : NTP_Time;
+        receive_ts     : NTP_Time;
+        transmit_ts    : NTP_Time;
+        extensions     : case (has_exts) of {
+				true  -> exts	: Extension_Field[] &until($input.length() > 24); 
+				false -> nil 	: empty;
+                         } &requires(has_exts);
+        mac_fields     : case (mac_len) of {
+                                20	-> mac		: NTP_MAC;
+                                24 	-> mac_ext 	: NTP_MAC_ext;
+                                default -> nil2 	: empty;
+                         } &requires(mac_len);
+} &let {
+        length          = sourcedata.length();
+	has_exts: bool	= (length - offsetof(extensions)) > 24;
+	mac_len: uint32	= (length - offsetof(mac_fields));
+} &byteorder=bigendian &exportsourcedata;
+
+# This format is for mode==6, control msg
+# See RFC 1119 for details
+type NTP_control_msg = record {
+        second_byte	: uint8;
+        sequence	: uint16;
+        status      	: uint16;    #TODO: this can be further parsed internally
+        association_id	: uint16;
+        offs		: uint16;
+        c	   	: uint16;
+	data		: bytestring &length=c;
+        mac_fields	: case (has_control_mac) of {
+                                true  -> mac : NTP_CONTROL_MAC;
+                                false -> nil : empty;
+                         } &requires(has_control_mac);
+} &let {
+        R	: bool   = (second_byte & 0x80) > 0;	# First bit of 8-bits value
+        E	: bool   = (second_byte & 0x40) > 0;	# Second bit of 8-bits value
+        M	: bool   = (second_byte & 0x20) > 0;	# Third bit of 8-bits value
+        OpCode	: uint8  = (second_byte & 0x1F);	# Last 5 bits of 8-bits value
+        length	= sourcedata.length();
+        has_control_mac: bool   = (length - offsetof(mac_fields)) == 12;
+} &byteorder=bigendian &exportsourcedata;
+
+# As in RFC 5905
+type NTP_MAC = record {
+	key_id	: uint32;
+	digest	: bytestring &length=16;
+} &length=20;
+
+# As in RFC 5906, same as NTP_MAC but with a 160 bit digest
+type NTP_MAC_ext = record {
+	key_id	: uint32;
+	digest	: bytestring &length=20;
+} &length=24;
+
+# As in RFC 1119
+type NTP_CONTROL_MAC = record {
+        key_id		: uint32;
+     	crypto_checksum	: bytestring &length=8;
+} &length=12;
+
+# As defined in RFC 5906
+type Extension_Field = record {
+  first_byte_ext: uint8;
```

Comment:
Indentation here and below should be tabs.

Reply:
ok